[SPARK-30456][SQL]  An exception to the clipping partition property causes "partition BETWEEN 'start' AND 'end' OR partition = 'other'" not to be clipped successfully

haixuan8192 opened the pull request:
https://issues.apache.org/jira/browse/SPARK-30456

PruneFileSourcePartitions PhysicalOperation pattern matching in class to predicate filter expressions segmentation, segmentation only 'AND' relationship, not considering 'OR' relationship;This effect can affect partition clipping, resulting in situations where clipping is not possible.For example: "partition BETWEEN 'start' AND 'end' OR (partition = 'other' AND column <> 'value')"

Commit:
An exception to the clipping partition property causes "partition BETWEEN 'start' AND 'end' OR partition = 'other'" not to be clipped successfully

Commit:
An exception to the clipping partition property causes "partition BETWEEN 'start' AND 'end' OR partition = 'other'" not to be clipped successfully

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

maropu commented:
hi, @haixuan8192, thanks for the work! Can you add tests before running the Jenkins? Also, can you describe more in the PR descritption?

haixuan8192 commented:
@maropu Sorry, there is a small problem in my local SBT environment, which is being solved now.When I'm querying a contiguous partition AND another partition (OR a partition), using "BETWEEN ~ AND ~ OR ~" won't cut successfully in the partition clipping phase because 'filters:Seq[Expression]' generated in PhysicalOperation 'doesn't take this 'OR' into account

Commit:
An exception to the clipping partition property causes "partition BETWEEN 'start' AND 'end' OR partition = 'other'" not to be clipped successfully

Commit:
Merge branch 'master' of https://github.com/apache/spark into haixuan8192

# Conflicts:
#	sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala

HyukjinKwon committed:
[SPARK-30480][PYTHON][TESTS] Increases the memory limit being tested in 'WorkerMemoryTest.test_memory_limit'

### What changes were proposed in this pull request?

This PR proposes to increase the memory in `WorkerMemoryTest.test_memory_limit` in order to make the test pass with PyPy.

The test is currently failed only in PyPy as below in some PRs unexpectedly:

```
Current mem limits: 18446744073709551615 of max 18446744073709551615

Setting mem limits to 1048576 of max 1048576

RPython traceback:
  File "pypy_module_pypyjit_interp_jit.c", line 289, in portal_5
  File "pypy_interpreter_pyopcode.c", line 3468, in handle_bytecode__AccessDirect_None
  File "pypy_interpreter_pyopcode.c", line 5558, in dispatch_bytecode__AccessDirect_None
out of memory: couldn't allocate the next arena
ERROR
```

It seems related to how PyPy allocates the memory and GC works PyPy-specifically. There seems nothing wrong in this configuration implementation itself in PySpark side.

I roughly tested in higher PyPy versions on Ubuntu (PyPy v7.3.0) and this test seems passing fine so I suspect this might be an issue in old PyPy behaviours.

The change only increases the limit so it would not affect actual memory allocations. It just needs to test if the limit is properly set in worker sides. For clarification, the memory is unlimited in the machine if not set.

### Why are the changes needed?

To make the tests pass and unblock other PRs.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Manually and Jenkins should test it out.

Closes #27186 from HyukjinKwon/SPARK-30480.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

Commit:
PruneFileSourcePartitions PhysicalOperation pattern matching in class to predicate filter expressions segmentation, segmentation only 'AND' relationship, not considering 'OR' relationship;This effect can affect partition clipping, resulting in situations where clipping is not possible.For example: "partition BETWEEN 'start' AND 'end' OR (partition = 'other' AND column <> 'value')"

Commit:
PruneFileSourcePartitions PhysicalOperation pattern matching in class to predicate filter expressions segmentation, segmentation only 'AND' relationship, not considering 'OR' relationship;This effect can affect partition clipping, resulting in situations where clipping is not possible.For example: "partition BETWEEN 'start' AND 'end' OR (partition = 'other' AND column <> 'value')"

Commit:
PruneFileSourcePartitions PhysicalOperation pattern matching in class to predicate filter expressions segmentation, segmentation only 'AND' relationship, not considering 'OR' relationship;This effect can affect partition clipping, resulting in situations where clipping is not possible.For example: "partition BETWEEN 'start' AND 'end' OR (partition = 'other' AND column <> 'value')"

Commit:
PruneFileSourcePartitions PhysicalOperation pattern matching in class to predicate filter expressions segmentation, segmentation only 'AND' relationship, not considering 'OR' relationship;This effect can affect partition clipping, resulting in situations where clipping is not possible.For example: "partition BETWEEN 'start' AND 'end' OR (partition = 'other' AND column <> 'value')"

Commit:
Merge remote-tracking branch 'origin/haixuan8192' into haixuan8192

# Conflicts:
#	sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala
